src/main/java/co/cask/publisher/PubSubOutputFormat.java

```diff
@@ -0,0 +1,207 @@
+  }
+
+  private RetrySettings getRetrySettings(int maxRetryTimeout) {
+    // only maxRetryTimeout is configurable by user, setting defaults for other parameters
+    return RetrySettings.newBuilder()
+      .setInitialRetryDelay(Duration.ofMillis(100))
+      .setRetryDelayMultiplier(2)
+      .setMaxRetryDelay(Duration.ofSeconds(2))
+      .setTotalTimeout(Duration.ofSeconds(maxRetryTimeout))
+      .setInitialRpcTimeout(Duration.ofSeconds(1))
+      .setMaxRpcTimeout(Duration.ofSeconds(5))
+      .build();
+  }
+
+  private BatchingSettings getBatchingSettings(long countSize, long bytesThreshold, long delayThreshold) {
+    BatchingSettings.Builder batchingSettingsBuilder = BatchingSettings.newBuilder();
+    batchingSettingsBuilder.setElementCountThreshold(countSize);
+    batchingSettingsBuilder.setRequestByteThreshold(bytesThreshold);
+    batchingSettingsBuilder.setDelayThreshold(Duration.ofMillis(delayThreshold));
+    return batchingSettingsBuilder.build();
+  }
+
+  /**
+   * Writer publishes messages to PubSub using the passed Publisher,
+   * batching and retrying are performed by Publisher based on how it is configured, If publishing a message fails
+   * we maintain error count and we throw exception when this error count exceeds a threshold.
+   */
+  public class PubSubRecordWriter extends RecordWriter<NullWritable, Text> {
+    private final Publisher publisher;
+    private final AtomicLong failures;
+    private final AtomicReference<Throwable> error;
+    private final long errorThrehold;
+
+    public PubSubRecordWriter(Publisher publisher, long errorThreshold) throws IOException {
+      this.publisher = publisher;
+      this.error = new AtomicReference<>();
+      this.errorThrehold = errorThreshold;
+      this.failures = new AtomicLong(0);
+    }
+
+    @Override
+    public void write(NullWritable key, Text value) throws IOException {
+      handleErrorIfAny();
+      ByteString data = ByteString.copyFromUtf8(value.toString());
+      PubsubMessage message = PubsubMessage.newBuilder().setData(data).build();
+      ApiFuture future = publisher.publish(message);
+      ApiFutures.addCallback(future, new ApiFutureCallback<String>() {
+        @Override
+        public void onFailure(Throwable throwable) {
+          error.set(throwable);
+          failures.incrementAndGet();
+        }
+
+        @Override
+        public void onSuccess(String s) {
+          // do nothing
+        }
+      });
+    }
+
+    private void handleErrorIfAny() throws IOException {
+      if (failures.get() > errorThrehold) {
+        throw new IOException(
+          String.format("Publishing to PubSub failed for %s records due to %s", failures.get(), error));
+      }
+    }
+
+    @Override
+    public void close(TaskAttemptContext taskAttemptContext) throws IOException {
+      try {
+        handleErrorIfAny();
+      } finally {
+        try {
+          publisher.shutdown();
```

Comment:
I think you have to block on futures that are not yet completed before shutting down, otherwise there could be unpublished message.

Reply:
it is handled as part of shutdown by the `Publisher`- https://github.com/GoogleCloudPlatform/google-cloud-java/blob/master/google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java#L406

Reply:
sorry, you meant the messages published by the output format, whereas the shutdown waits for the published outstanding messages.